app-tests/run.sh

```diff
@@ -0,0 +1,150 @@
+  cd opal-tests-policy-repo
+  git checkout -b $POLICY_REPO_BRANCH
+  git push --set-upstream origin $POLICY_REPO_BRANCH
+  cd -
+
+  # That's for the docker-compose to use, set ssh key from "~/.ssh/id_rsa", unless another path/key data was configured
+  export OPAL_POLICY_REPO_SSH_KEY
+  OPAL_POLICY_REPO_SSH_KEY_PATH=${OPAL_POLICY_REPO_SSH_KEY_PATH:-~/.ssh/id_rsa}
+  OPAL_POLICY_REPO_SSH_KEY=${OPAL_POLICY_REPO_SSH_KEY:-$(cat "$OPAL_POLICY_REPO_SSH_KEY_PATH")}
+}
+
+# Setup
+generate_opal_keys
+prepare_policy_repo
+
+
+function compose {
+  docker compose -f ./docker-compose-app-tests.yml --env-file .env "$@"
+}
+
+function check_clients_logged {
+  echo "- Looking for msg '$1' in client's logs"
+  compose logs --index 1 opal_client | grep -q "$1"
+  compose logs --index 2 opal_client | grep -q "$1"
+}
+
+function check_no_error {
+  # Without index would output all replicas
+  if compose logs opal_client | grep -q 'ERROR'; then
+    echo "- Found error in logs"
+    exit 1
+  fi
+}
+
+function clean_up {
+    ARG=$?
+    if [[ "$ARG" -ne 0 ]]; then
+      echo "*** Test Failed ***"
+      echo ""
+      compose logs
+    else
+      echo "*** Test Passed ***"
+      echo ""
+    fi
+    compose down
+    cd opal-tests-policy-repo; git push -d origin $POLICY_REPO_BRANCH; cd - # Remove remote tests branch
+    rm -rf ./opal-tests-policy-repo
+    exit $ARG
+}
+trap clean_up EXIT
+
+# Bring up OPAL containers
+compose down --remove-orphans
+compose up -d
+sleep 10
+
+# Check containers started correctly
+check_clients_logged "Connected to PubSub server"
+check_clients_logged "Got policy bundle"
+check_clients_logged 'PUT /v1/data/static -> 204'
+check_no_error
+
+function test_push_policy {
+  echo "- Testing pushing policy $1"
+  regofile="$1.rego"
+  cd opal-tests-policy-repo
+  echo "package $1" > "$regofile"
+  git add "$regofile"
+  git commit -m "Add $regofile"
+  git push
+  cd -
+
+  curl -s --request POST 'http://localhost:7002/webhook' --header 'Content-Type: application/json' --header 'x-webhook-token: xxxxx' --data-raw '{"gitEvent":"git.push","repository":{"git_url":"git@github.com:permitio/opal-tests-policy-repo.git"}}'
+  sleep 5
+  check_clients_logged "PUT /v1/policies/$regofile -> 200"
+}
+
+function test_data_publish {
+  echo "- Testing data publish for user $1"
+  user=$1
+  OPAL_CLIENT_TOKEN=$OPAL_DATA_SOURCE_TOKEN opal-client publish-data-update --src-url https://api.country.is/23.54.6.78 -t policy_data --dst-path "/users/$user/location"
+  sleep 5
+  check_clients_logged "PUT /v1/data/users/$user/location -> 204"
+}
+
+function test_statistics {
+    echo "- Testing statistics feature"
+    # Make sure 2 servers & 2 clients (repeat few times cause different workers might response)
+    for _ in {1..10}; do
+      curl -s 'http://localhost:7002/stats' --header "Authorization: Bearer $OPAL_DATA_SOURCE_TOKEN" | grep '"client_count":2,"server_count":2'
+    done
+}
+
+test_data_publish "bob"
```

Comment:
would make some sort of main function here to understand what happens when, it is currently hard to understand as the functions and global scope are mixed together